fix: wau or mau with week or month or longer intervals falls back to dau

aspicer opened the pull request:
## Problem

We get a healthy number of support issues from people being confused about primarily WAU logic for weekly intervals.

WAUs don't really exist for intervals of a week or longer. Our current query returns something which is wrong, which is different than what the actors modal shows.

Make our query consistent with the actors modal. In the case of a week or longer interval, WAUs just because unique users in the query.

## Changes

TrendsQueryRunner silently changes weekly or monthly math to unique user math if the interval is longer.

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Wrote tests!

aspicer committed:
waumau

aspicer committed:
active

aspicer committed:
test

aspicer committed:
test

aspicer committed:
clean up

aspicer committed:
comment

aspicer committed:
tooltips

github-actions[bot] committed:
Update query snapshots

aspicer committed:
fix typing

aspicer committed:
Merge branch 'aspicer/waumau' of github.com:PostHog/posthog into aspicer/waumau

aspicer committed:
types

aspicer committed:
Merge remote-tracking branch 'origin/master' into aspicer/waumau

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

aspicer committed:
stickiness

aspicer committed:
Merge branch 'aspicer/waumau' of github.com:PostHog/posthog into aspicer/waumau

aspicer committed:
tests

thmsobrmlr commented on posthog/hogql_queries/insights/utils/utils.py:
```diff
@@ -15,3 +19,32 @@ def get_start_of_interval_hogql(interval: str, *, team: Team, source: Optional[a
     return f"{trunc_func}({source}{f', {int((WeekStartDay(team.week_start_day or 0)).clickhouse_mode)}' if trunc_func == 'toStartOfWeek' else ''})"
+
+
+def convert_active_user_math_based_on_interval(query: T) -> T:
```
Not sure if it makes sense in this case, but thought it's worth pointing out that we have a `serialize_query` method in `posthog/schema_helpers.py` thats responsible for transforming queries into a canonical representation. This is useful because then semantically equal queries with different syntax still use the same cache key.

It's a minor optimization, but could have unintended side effects, e.g. updating the query configuration and causing more confusion for the user, so maybe not that useful here.

aspicer replied:
Thanks - i modified this a bit and added that functionality to both the FE and the BE - do you mind taking a quick look?

thmsobrmlr reviewed (approved):
👍  for functional correctness. 

Wondering if we should inform users about this change in some way though, as this could cause confusion. Could be an email to all users who have such an insight or a frontend side hint on affected insights.

Also should we forbid this configuration frontend side or at least hint that this is problematic?

thmsobrmlr reviewed (approved):
👍  for functional correctness. 

Wondering if we should inform users about this change in some way though, as this could cause confusion. Could be an email to all users who have such an insight or a frontend side hint on affected insights.

Also should we forbid this configuration frontend side or at least hint that this is problematic?

aspicer committed:
ah

aspicer commented:
> 👍 for functional correctness.
> 
> Wondering if we should inform users about this change in some way though, as this could cause confusion. Could be an email to all users who have such an insight or a frontend side hint on affected insights.
> 
> Also should we forbid this configuration frontend side or at least hint that this is problematic?

Thanks. Great suggestion. Digging into this made me realize we had a bug in the code that was stoping these tooltips from appearing for Total Count logic, which warned about the same thing. The bug was also causing these math types to appear in stickiness where it wasn't supported. Fixed those and added a warning!
<img width="908" alt="image" src="https://github.com/user-attachments/assets/b7ddab27-9302-4223-91c9-2ab0e3291d2e" />

aspicer committed:
fix

aspicer committed:
fix

aspicer committed:
clean

aspicer committed:
merge

aspicer committed:
clean up types

aspicer committed:
updateS

aspicer committed:
weird stuff

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

aspicer committed:
test

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

aspicer committed:
query utils

aspicer committed:
Merge branch 'aspicer/waumau' of github.com:PostHog/posthog into aspicer/waumau

aspicer committed:
types

thmsobrmlr reviewed (approved):
Thanks for the follow-up, that looks good!
